### PR TITLE
QALI-489: Replace the micrometer Meter registry by our own implementa…

### PR DESCRIPTION
**Pull request by oceenachi**
…tion

### CODE REVIEW

**Review by ericjesse — CHANGES_REQUESTED** [review]

**Review by ericjesse — CHANGES_REQUESTED** [review]

File: influxdb/src/main/kotlin/io/qalipsis/plugins/influxdb/config/InfluxDbMeasurementPublisherFactory.kt [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+    override fun getPublisher(): InfluxdbMeasurementPublisher {
+        return InfluxdbMeasurementPublisher(
+            configuration = configuration,
+            coroutineScope = coroutineScope
+        )
+    }
+
+    companion object {
+
+        private const val INFLUXDB_CONFIGURATION = "${MetersConfig.EXPORT_CONFIGURATION}.influxdb"
```
Comment (ericjesse) [inline influxdb/src/main/kotlin/io/qalipsis/plugins/influxdb/config/InfluxDbMeasurementPublisherFactory.kt:55]: Move the constants to the config class.

File: influxdb/src/main/kotlin/io/qalipsis/plugins/influxdb/config/InfluxDbMeasurementPublisherFactory.kt [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+import io.qalipsis.plugins.influxdb.meters.InfluxDbMeasurementConfiguration
+import io.qalipsis.plugins.influxdb.meters.InfluxdbMeasurementPublisher
+import jakarta.inject.Named
+import jakarta.inject.Singleton
+import kotlinx.coroutines.CoroutineScope
+
+/**
+ * Configuration for the export of QALIPSIS meters to InfluxDb.
+ *
+ * @author Palina Bril
```
Comment (ericjesse) [inline influxdb/src/main/kotlin/io/qalipsis/plugins/influxdb/config/InfluxDbMeasurementPublisherFactory.kt:34]: Change the author here.

File: influxdb/src/main/kotlin/io/qalipsis/plugins/influxdb/config/InfluxDbMeasurementPublisherFactory.kt [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+import io.qalipsis.api.config.MetersConfig
+import io.qalipsis.api.meters.MeasurementPublisherFactory
+import io.qalipsis.plugins.influxdb.meters.InfluxDbMeasurementConfiguration
+import io.qalipsis.plugins.influxdb.meters.InfluxdbMeasurementPublisher
+import jakarta.inject.Named
+import jakarta.inject.Singleton
+import kotlinx.coroutines.CoroutineScope
+
+/**
+ * Configuration for the export of QALIPSIS meters to InfluxDb.
```
Comment (ericjesse) [inline influxdb/src/main/kotlin/io/qalipsis/plugins/influxdb/config/InfluxDbMeasurementPublisherFactory.kt:32]: The documentation is not accurate.

File: influxdb/src/main/kotlin/io/qalipsis/plugins/influxdb/meters/InfluxdbMeasurementPublisher.kt [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+            )
+        }"
+    }
+
+
+    private suspend fun saveRecords(records: Collection<String>) {
+        try {
+            writeClient.writeRecords(records, WritePrecision.MS, configuration.bucket, configuration.org)
+            logger.debug { "Successfully sent ${records.size} metrics to InfluxDb" }
+            logger.trace { "onSuccess totally processed" }
```
Comment (ericjesse) [inline influxdb/src/main/kotlin/io/qalipsis/plugins/influxdb/meters/InfluxdbMeasurementPublisher.kt:135]: Do you really need both traces?

File: influxdb/src/main/kotlin/io/qalipsis/plugins/influxdb/meters/InfluxdbMeasurementPublisher.kt [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+        } catch (ex: Exception) {
+            logger.warn(ex) { "${records.size} could not be save: ${ex.message}" }
+            throw ex
+        }
+    }
+
+    override suspend fun stop() {
+        logger.debug { "Stopping the meter publication of meters" }
+        runBlocking(coroutineScope.coroutineContext) {
```
Comment (ericjesse) [inline influxdb/src/main/kotlin/io/qalipsis/plugins/influxdb/meters/InfluxdbMeasurementPublisher.kt:144]: Why are you using a runblocking here? You are already in a suspended context.

File: influxdb/src/main/kotlin/io/qalipsis/plugins/influxdb/meters/InfluxdbMeasurementPublisher.kt [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+            publicationLatch.await()
+        }
+        logger.debug { "Closing the InfluxDbKotlin client" }
+        tryAndLogOrNull(logger) {
+            client.close()
+        }
+        logger.debug { "The meters publication logger was stopped" }
+    }
+
+    private fun String.replaceSpaces() = this.replace(" ", "-").lowercase()
```
Comment (ericjesse) [inline influxdb/src/main/kotlin/io/qalipsis/plugins/influxdb/meters/InfluxdbMeasurementPublisher.kt:155]: The function does more than replacing spaces. `format` or `sanitize()` would probably be better.

File: influxdb/src/main/resources/application-influxdb.yml [hunk 7]
```diff
@@ -20,7 +20,6 @@ meters:
     influxdb:
       url: http://localhost:8086
       enabled: false
-      step: PT10S
       org: qalipsis
       bucket: qalipsis-meter
-
+      prefix: influxdb
```
Comment (ericjesse) [inline influxdb/src/main/resources/application-influxdb.yml:25]: why not using `qalipsis` as prefix?

File: influxdb/src/main/kotlin/io/qalipsis/plugins/influxdb/meters/InfluxdbMeasurementPublisher.kt [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+                it.statistic.value to it.value
+            } else {
+                "${it.statistic.value}_${it.observationPoint}" to it.value
+            }
+        }
+        val tags = if (meterId.tags.isNotEmpty()) {
+            meterId.tags.mapValues { (_, value) -> "\"$value\"" }
+                .entries.joinToString(",") + ",metric_type=${meterId.type.value}"
+        } else "metric_type=${meterId.type.value}"
+        return "${configuration.prefix}.${meterId.campaignKey.replaceSpaces()}.${meterId.scenarioName.replaceSpaces()}.${meterId.stepName.replaceSpaces()}.${meterId.meterName.replaceSpaces()},$tags ${
```
Comment (ericjesse) [inline influxdb/src/main/kotlin/io/qalipsis/plugins/influxdb/meters/InfluxdbMeasurementPublisher.kt:123]: What if the prefix is blank, will the name start with a dot?
I would rather calculate the complete prefix in the init {} block to determine whether it is a totally empty string or a real prefix with a dot.

**Review by ericjesse — CHANGES_REQUESTED** [review]

File: influxdb/src/main/resources/application-influxdb.yml (hunk 7 above)
Comment (oceenachi) [reply influxdb/src/main/resources/application-influxdb.yml:25]: I thought the default prefix for the plugins should be the plugin name though.

Comment (ericjesse) [reply influxdb/src/main/resources/application-influxdb.yml]: No, because other plugins will never store data in the influxdb.

Comment (oceenachi) [reply influxdb/src/main/resources/application-influxdb.yml]: Noted

File: influxdb/src/main/kotlin/io/qalipsis/plugins/influxdb/meters/InfluxdbMeasurementPublisher.kt [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+
+    private lateinit var client: InfluxDBClientKotlin
+
+    private lateinit var writeClient: WriteKotlinApi
+
+    private lateinit var publicationLatch: SuspendedCountLatch
+
+    private lateinit var publicationSemaphore: Semaphore
+
+    private var prefix: String = if (configuration.prefix.isNotEmpty()) "${configuration.prefix}." else ""
```
Comment (ericjesse) [inline influxdb/src/main/kotlin/io/qalipsis/plugins/influxdb/meters/InfluxdbMeasurementPublisher.kt:57]: why var? val should be enough

File: influxdb/src/main/kotlin/io/qalipsis/plugins/influxdb/meters/InfluxdbMeasurementPublisher.kt [hunk 10]
```diff
@@ -0,0 +1,158 @@
...
+import kotlinx.coroutines.sync.Semaphore
+import kotlinx.coroutines.sync.withPermit
+
+/**
+ * Implementation of measurement publisher to export meters to influxdb.
+ *
+ * @author Francisca Eze
+ */
+@Requires(beans = [InfluxDbMeasurementConfiguration::class])
+class InfluxdbMeasurementPublisher(
```
Comment (ericjesse) [inline influxdb/src/main/kotlin/io/qalipsis/plugins/influxdb/meters/InfluxdbMeasurementPublisher.kt:44]: No need for the requires, the publisher is not a bean. It is created by the factory.

File: influxdb/src/main/kotlin/io/qalipsis/plugins/influxdb/meters/InfluxDbMeasurementConfiguration.kt [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+ * @property url string URL to the Influxdb instance.
+ * @property publishers number of concurrent publication of events that can be run defaults to 1 (no concurrency).
+ * @property userName name of the user to use for basic authentication when connecting to InfluxDb.
+ * @property password password of the user to use for basic authentication when connecting to InfluxDb.
+ * @property org name of the organization to use for basic authentication when connecting to InfluxDb.
+ * @property bucket password of the bucket to use for saving events to InfluxDb.
+ * @property prefix property prefix to prepend to configuration names, defaults to qalipsis.
+ *
+ */
+@Requires(property = "$INFLUXDB_CONFIGURATION.enabled", value = "true")
```
Comment (ericjesse) [inline influxdb/src/main/kotlin/io/qalipsis/plugins/influxdb/meters/InfluxDbMeasurementConfiguration.kt:39]:
```suggestion
@Requires(property = "$INFLUXDB_ENABLED", value = "StringUtils.TRUE")
```

**Review by ericjesse — APPROVED** [review]

**Pull request merged**